flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/deduplicate/DeduplicateFunctionHelper.java

```diff
@@ -96,12 +109,20 @@ static void processLastRowOnChangelog(
                 currentRow.setRowKind(RowKind.INSERT);
                 out.collect(currentRow);
             } else {
-                if (generateUpdateBefore) {
-                    preRow.setRowKind(RowKind.UPDATE_BEFORE);
-                    out.collect(preRow);
+                if (!isStateTtlEnabled && equaliser.equals(preRow, currentRow)) {
+                    // currentRow is the same as preRow and state cleaning is not enabled.
+                    // We do not emit retraction and update message.
+                    // If state cleaning is enabled, we have to emit messages to prevent too early
+                    // state eviction of downstream operators.
```

Comment:
Do we need to take care `isStateTtlEnabled ` here ? The `preRow`  value who comes from `state` has been influenced by `isStateTtlEnabled`.

Reply:
We may need to enable this optimization only when state ttl is disabled to prevent too early state eviction of downstream operators. 
just like the implementation in 
https://github.com/apache/flink/blob/f3db4220f5c8730e065734cff16237c7743b390f/flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/aggregate/GroupAggFunction.java#L170

Reply:
Got it, make sense to me